File: CCC/searchingForStrings.cpp
#include <iostream>
#include <string>
#include <map>

using namespace std;

bool permutation(map<char,int> m1, map<char,int> m2){
    for (int i = 0; i < 26; i++)
    {
        char test = 'a'+i;
        if (m1[test] != m2[test]){
            return false;
        }
    }
    return true;
}

int main(){
    string needle, haystack;
    int ans =0;
    cin>>needle;
    cin>>haystack;
    map<string,bool> tested;
    map<char,int> needle_chars, haystack_chars;
    for(int i=0;i<needle.size();i++){
        needle_chars[needle[i]]++;
    }

    for(int i=0;i<haystack.size()-needle.size()+1;i++) {
        string sub_str = haystack.substr(i,needle.size());
        if (i==0)
        {
            for(int j=0;j<sub_str.size();j++)
            {
                haystack_chars[sub_str[j]]++;
            }
        }
        else
        {
            char added = haystack[i+needle.size()-1];
            char removed = haystack[i-1];
            haystack_chars[added]++;
            haystack_chars[removed]--;                        
        }
        if (permutation(needle_chars, haystack_chars) && !tested[sub_str])
            {
                ans++;
                tested[sub_str] = true;
            }
    }

    cout<<ans<<endl;
}

File: CCC/searchingForStrings.py
def is_perm(m1,m2):
    for i in range(26):
        letter = 'a'+i
        if m1[letter] != m2[letter]:
            return False
    return True
    
needle = input()
haystack = input()

needle_f = {}
haystack_f = {}
used = {}

for char in needle:
    needle_f[char] +=1

ans =0

for i in range(0, len(haystack)-len(needle)+1):
    sub_str = haystack[i:i+len(needle)]
    if i == 0:
        for j in sub_str:
            haystack_f[j] +=1
    else:
        added = haystack[i+len(needle)-1]
        removed = haystack[i-1]
        haystack_f[added] +=1
        haystack_f[removed] -=1
        if(is_perm(haystack_f, needle_f) and used[sub_str] == False):
            ans +=1
            used[sub_str] = True

print(ans)